### PR TITLE
Support Labels

### PR DESCRIPTION
**Pull request by fahedouch**
- [x]  support labels for networks
- [x]  support labels for volumes 
- [x]  support labels for compose resources (networks and volumes)
- [x]  add labels to volume inspect elements
- [x]  add labels to network inspect elements

related issue https://github.com/containerd/nerdctl/issues/258

### CODE REVIEW

File: pkg/netutil/netutil_linux.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ const ConfigListTemplate = `{
   "cniVersion": "0.4.0",
   "name": "{{.Name}}",
   "nerdctlID": {{.ID}},
+  "NerdctlLabels": {{.Labels}},
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil_linux.go:33]: N -> n

File: pkg/mountutil/mountutil.go [hunk 2]
```diff
@@ -54,7 +54,7 @@ func ProcessFlagV(s string, volStore volumestore.VolumeStore) (*Processed, error
 		dst = s
 		res.AnonymousVolume = idgen.GenerateID()
 		logrus.Debugf("creating anonymous volume %q, for %q", res.AnonymousVolume, s)
-		anonVol, err := volStore.Create(res.AnonymousVolume)
+		anonVol, err := volStore.Create(res.AnonymousVolume, []string{})
```
Comment (AkihiroSuda) [inline pkg/mountutil/mountutil.go:57]: Compilation is failing, not sure why though

```
#15 61.55 # github.com/containerd/nerdctl/pkg/mountutil
#15 61.55 pkg/mountutil/mountutil.go:72:32: not enough arguments in call to volStore.Create
#15 61.55 	have (string)
#15 61.55 	want (string, []string)
```
https://github.com/containerd/nerdctl/pull/274/checks?check_run_id=2984458522#step:3:1290

File: pkg/mountutil/volumestore/volumestore.go [hunk 3]
```diff
@@ -55,9 +57,11 @@ func New(dataStore, ns string) (VolumeStore, error) {
 // DataDirName is "_data"
 const DataDirName = "_data"
 
+const volumeFileName = "volume.json"
```
Comment (AkihiroSuda) [inline pkg/mountutil/volumestore/volumestore.go:60]: const `volumeJSONFileName`, or `volumeMetaFileName`

File: pkg/mountutil/volumestore/volumestore.go [hunk 4]
```diff
@@ -87,6 +91,19 @@ func (vs *volumeStore) Create(name string) (*native.Volume, error) {
 		if err := os.Mkdir(volDataPath, 0755); err != nil {
 			return err
 		}
+
+		volFilePath := filepath.Join(volPath, volumeFileName)
+		labelsMap := strutil.ConvertKVStringsToMap(labels)
+
+		volumeOpts := make(map[string]interface{})
+		volumeOpts["nerdctlLabels"] = labelsMap
```
Comment (AkihiroSuda) [inline pkg/mountutil/volumestore/volumestore.go:99]: Just `labels` is fine for volumes, as we own the entire JSON by ourselves

Comment (AkihiroSuda) [reply pkg/mountutil/volumestore/volumestore.go:99]: And better to define a struct

File: pkg/mountutil/volumestore/volumestore.go [hunk 5]
```diff
@@ -155,3 +184,14 @@ func (vs *volumeStore) Remove(names []string) ([]string, error) {
 	err := lockutil.WithDirLock(vs.dir, fn)
 	return removed, err
 }
+
+func NerdctlLabels(b []byte) *map[string]string {
+	type nerdctlConfigList struct {
+		NerdctlLabels *map[string]string `json:"nerdctlLabels,omitempty"`
```
Comment (AkihiroSuda) [inline pkg/mountutil/volumestore/volumestore.go:190]: same as above

File: pkg/netutil/netutil.go [hunk 6]
```diff
@@ -45,12 +47,13 @@ type CNIEnv struct {
...
 
 func DefaultConfigList(e *CNIEnv) (*NetworkConfigList, error) {
-	return GenerateConfigList(e, DefaultID, DefaultNetworkName, DefaultCIDR)
+	return GenerateConfigList(e, nil, DefaultID, DefaultNetworkName, DefaultCIDR)
 }
 
 type ConfigListTemplateOpts struct {
 	ID           int
 	Name         string // e.g. "nerdctl"
+	Labels       string // e.g. `,{"group":"bigger"}`
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil.go:56]: What does bigger mean?

Comment (AkihiroSuda) [inline pkg/netutil/netutil.go:56]: no `,` should be needed

File: pkg/mountutil/mountutil.go (hunk 2 above)
Comment (AkihiroSuda) [reply pkg/mountutil/mountutil.go:57]: @ktock @fuweid Could you take a look?

I don't see `volstore.Create()` in `mountutil.go:72`, not sure why this is failing

Comment (ktock) [reply pkg/mountutil/mountutil.go:57]: @fahedouch This PR needs rebase.

https://github.com/containerd/nerdctl/blob/fde3eae7cc21f6b010e683026837cab11d544f9f/pkg/mountutil/mountutil.go#L72

File: pkg/netutil/netutil.go (hunk 6 above)
Comment (fahedouch) [reply pkg/netutil/netutil.go]: it was just an example, I changed to a more meaningful example

Comment (fahedouch) [reply pkg/netutil/netutil.go]: it was just an example, I changed to a more meaningful example

Comment (fahedouch) [reply pkg/netutil/netutil.go]: fixed

File: pkg/mountutil/volumestore/volumestore.go (hunk 5 above)
Comment (fahedouch) [reply pkg/mountutil/volumestore/volumestore.go]: fixed

File: pkg/mountutil/volumestore/volumestore.go (hunk 4 above)
Comment (fahedouch) [reply pkg/mountutil/volumestore/volumestore.go]: fixed

File: pkg/mountutil/volumestore/volumestore.go (hunk 3 above)
Comment (fahedouch) [reply pkg/mountutil/volumestore/volumestore.go]: fixed

File: pkg/mountutil/mountutil.go (hunk 2 above)
Comment (fahedouch) [reply pkg/mountutil/mountutil.go:58]: yep thank you @ktock

File: pkg/netutil/netutil_windows.go [hunk 7]
```diff
@@ -31,6 +31,7 @@ const ConfigListTemplate = `{
   "cniVersion": "0.4.0",
   "name": "{{.Name}}",
   "nerdctlID": {{.ID}},
+  "nerdctlLabels": "{{.Labels}}",
```
Comment (AkihiroSuda) [inline pkg/netutil/netutil_windows.go:34]: No quotation

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**